### PR TITLE
DCCP Socket

### PR DESCRIPTION
**Pull request by yuyan0**
I am still working on pull(), but you could merge these first when you have time :-)

### CODE REVIEW

**Comment by yilongli** [comment]
There are many changes that are not related to the `DCCPSocket` in this pull request. And there are several places in the code that I do not quite understand. I will leave my comments here and you can continue to work in your own fork until we resolve them to make sure the change doesn't break anything that works before.

File: src/common.hh [hunk 1]
```diff
@@ -102,7 +101,7 @@ generateRandom()
     return std::rand();
 }
 
-bool poll(UDPSocket* socket, UDPSocket::received_datagram& datagram)
+bool poll(DCCPSocket* socket, char*& datagram)
```
Comment (yilongli) [inline src/common.hh]: why change the type of `datagram`?

File: src/wire_format.hh [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 namespace WireFormat {
 
-enum class Opcode : uint8_t {
+enum Opcode : uint8_t {
```
Comment (yilongli) [inline src/wire_format.hh]: why change `enum class` back to `enum`? It's not really related to DCCPSocket.

File: src/socket.hh [hunk 3]
```diff
@@ -85,4 +85,29 @@ public:
...
+  void listen( const int backlog = 16 );
+
+  /* accept a new incoming connection */
+  DCCPSocket accept( void );
+
+  /* send datagram to connected address */
+  void send( const char* payload, int payload_len );
+
+  /* receive datagram from connected address */
+  char* recv( void );
```
Comment (yilongli) [inline src/socket.hh]: Now I see why you have to change the signature of `poll` in common.cc. It's better to implement a DCCPSocket::received_datagram like the UDPSocket return that here.

File: src/receiver.cc [hunk 4]
```diff
@@ -36,83 +41,101 @@ int checkArgs(int argc, char *argv[]) {
...
+        printUsage(argv[0]);
         return -1;
     }
-    return debug_f;
+
+    return 0;
 }
 
-int main( int argc, char *argv[] )
+DCCPSocket* respondHandshake(Tub<WireFormat::HandshakeReq>& req)
```
Comment (yilongli) [inline src/receiver.cc]: why remove the `unique_ptr`? I put them there to avoid having to free pointers manually.

File: src/receiver.cc [hunk 5]
```diff
@@ -152,47 +176,61 @@ int main( int argc, char *argv[] )
...
+    RaptorQDecoder decoder(req->otiCommon, req->otiScheme);
+    size_t decoderPaddedSize = 0;
+    for (int i = 0; i < decoder.blocks(); i++) {
+        decoderPaddedSize += decoder.block_size(i);
     }
 
+    // Create the receiving file
+    int fd = SystemCall("open the file to be written",
+    //       open(req->fileName, O_RDWR | O_CREAT | O_TRUNC, (mode_t)0600));
+             open("demo.out", O_RDWR | O_CREAT | O_TRUNC, (mode_t)0600));
```
Comment (yilongli) [inline src/receiver.cc]: why "demo.out" and not `req->fileName`?

File: src/sender.cc [hunk 6]
```diff
@@ -53,45 +92,32 @@ std::unique_ptr<UDPSocket> initiateHandshake(const RaptorQEncoder& encoder,
...
+void sendSymbol(DCCPSocket *socket,
                 RaptorQSymbolIterator &symbolIterator)
 {
     static RaptorQSymbol symbol {0};
     auto begin = symbol.begin();
     (*symbolIterator)(begin, symbol.end());
 
-    // Wait until the congestion controller gives us a pass
-    congestionControl();
+    usleep(5000);
```
Comment (yilongli) [inline src/sender.cc]: why sleep anymore since we have the DCCPSocket doing congestion control for us?

File: src/receiver.cc [hunk 7]
```diff
@@ -36,83 +41,101 @@ int checkArgs(int argc, char *argv[]) {
...
+        free(datagram); 
+
         uint8_t sbn = downCast<uint8_t>(dataPacket->id >> 24);
         uint32_t esi = (dataPacket->id << 8) >> 8;
 
-        if (debug_f) {
-            printf("Received sbn = %u, esi = %u\n", static_cast<uint32_t>(sbn),
-                   esi);
+        if (DEBUG_F) {
+        //printf("Received sbn = %u, esi = %u\n", static_cast<uint32_t>(sbn), esi);
```
Comment (yilongli) [inline src/receiver.cc]: why comment it out? It's pretty useful for debugging.

File: src/receiver.cc [hunk 8]
```diff
@@ -36,83 +41,101 @@ int checkArgs(int argc, char *argv[]) {
...
     uint32_t maxSymbolRecv[MAX_BLOCKS] {0};
     uint32_t repairSymbolInterval = INIT_REPAIR_SYMBOL_INTERVAL;
-    while (decodedBlocks.count() < decoder.blocks()) {
+
+    char* datagram;
+
+    while (1) {
         datagram = socket->recv();
-        Tub<WireFormat::DataPacket> dataPacket(datagram.payload);
+        // sender has closed the connection
```
Comment (yilongli) [inline src/receiver.cc]: I see that you remove the "teardown" phase. Are you certain that for a datagram socket (not stream socket),  when the sender shutdowns, the `recv` method on the receiver side will return 0?

File: src/wire_format.hh (hunk 2 above)
Comment (yuyan0) [inline src/wire_format.hh]: Just want to use the values, say ACK, directly in the namespace WireFormat.

File: src/socket.hh (hunk 3 above)
Comment (yuyan0) [inline src/socket.hh]: Actually it's not necessary to return received_datagram. We don't need the recv_len or peer address any more. Note that after calling connect(), we can directly use send() or recv().

File: src/receiver.cc (hunk 4 above)
Comment (yuyan0) [inline src/receiver.cc]: Yes, we can have that. I free the socket manually without introducing any error though.

File: src/receiver.cc (hunk 5 above)
Comment (yuyan0) [inline src/receiver.cc]: For debugging now.

File: src/receiver.cc (hunk 7 above)
Comment (yuyan0) [inline src/receiver.cc]: For my own debugging poll(). Too many outputs with this line. Let's discuss debugging after fixing other normal function.

**Pull request closed without merging**